src/codecs/openexr.rs

```diff
@@ -0,0 +1,495 @@
+        // therefore we currently read all lines at once
+        // Todo: optimize for specific exr.line_order?
+       self.total_bytes()
+    }
+
+    // reads with or without alpha, depending on `self.alpha_preference` and `self.alpha_present_in_file`
+    fn read_image_with_progress<F: Fn(Progress)>(self, unaligned_bytes: &mut [u8], progress_callback: F) -> ImageResult<()> {
+        let blocks_in_header = self.selected_exr_header().chunk_count as u64;
+        let channel_count = self.color_type().channel_count() as usize;
+
+        let display_window = self.selected_exr_header().shared_attributes.display_window;
+        let data_window_offset = self.selected_exr_header().own_attributes.layer_position - display_window.position;
+
+        let result = read()
+            .no_deep_data().largest_resolution_level()
+            .rgba_channels(
+                move |_size, _channels| {
+                    vec![0_f32; display_window.size.area() * channel_count]
+                },
+
+                move |buffer, index_in_data_window, (r,g,b,a_or_1): (f32,f32,f32,f32)| {
+                    let index_in_display_window = index_in_data_window.to_i32() + data_window_offset;
+
+                    // only keep pixels inside the data window
+                    // TODO filter chunks based on this
+                    if index_in_display_window.x() >= 0 && index_in_display_window.y() >= 0
+                        && index_in_display_window.x() < display_window.size.width() as i32
+                        && index_in_display_window.y() < display_window.size.height() as i32
+                    {
+                        let index_in_display_window = index_in_display_window.to_usize("index bug").unwrap();
+                        let first_f32_index = index_in_display_window.flat_index_for_size(display_window.size);
+
+                        buffer[first_f32_index * channel_count .. (first_f32_index + 1) * channel_count]
+                            .copy_from_slice(&[r, g, b, a_or_1][0 .. channel_count]);
+
+                        // TODO white point chromaticities + srgb/linear conversion?
+                    }
+                }
+            )
+            .first_valid_layer() // TODO select exact layer by self.header_index?
+            .all_attributes()
+            .on_progress(|progress| {
+                progress_callback(Progress::new(
+                    (progress*blocks_in_header as f64) as u64, blocks_in_header) // TODO precision errors?
+                );
+            })
+            .from_chunks(self.exr_reader).map_err(to_image_err)?;
+
+        // TODO this copy is strictly not necessary, but the exr api is a little too simple for reading into a borrowed target slice
+
+        // this cast is safe and works with any alignment, as bytes are copied, and not f32 values.
+        unaligned_bytes.copy_from_slice(bytemuck::cast_slice(result.layer_data.channel_data.pixels.as_slice()));
```

Comment:
This copy operation will panic if the caller passes a buffer with invalid length. We should probably check this and then return an error bofore calling `copy_from_slice`, right?